From ac7f91b5c8229361a693f6dea373f4e63483fe23 Mon Sep 17 00:00:00 2001
From: Jeroen Hofstee <jhofstee@victronenergy.com>
Date: Tue, 3 Dec 2024 11:33:30 +0100
Subject: [PATCH 5/8] can: dev: better handle do_set_mode(dev, CAN_MODE_START)
 failing

Set the carrier to on before calling set_mode and turn it off, if an
error occured. Upstream commit 6841cab8c4504 ("can: dev: can_restart():
fix race condition between controller restart and netif_carrier_on()")
roughly does the same. This unconditionally increments restarts, while
in upstream they changed that behaviour later on.
---
 drivers/net/can/dev/dev.c | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/drivers/net/can/dev/dev.c b/drivers/net/can/dev/dev.c
index aac91420e264..2726d40035a3 100644
--- a/drivers/net/can/dev/dev.c
+++ b/drivers/net/can/dev/dev.c
@@ -599,15 +599,16 @@ static void can_restart(struct net_device *dev)
 	netif_rx_ni(skb);
 
 restart:
-	netdev_dbg(dev, "restarted\n");
-	priv->can_stats.restarts++;
-
 	/* Now restart the device */
-	err = priv->do_set_mode(dev, CAN_MODE_START);
-
+	priv->can_stats.restarts++;
 	netif_carrier_on(dev);
-	if (err)
-		netdev_err(dev, "Error %d during restart", err);
+	err = priv->do_set_mode(dev, CAN_MODE_START);
+	if (err) {
+		netdev_err(dev, "Restart failed, error %pe\n", ERR_PTR(err));
+		netif_carrier_off(dev);
+	} else {
+		netdev_dbg(dev, "Restarted\n");
+	}
 }
 
 static void can_restart_work(struct work_struct *work)
-- 
2.43.0

